### DISCUSSION TITLE
User installed packages when shipping composer.lock

### DISCUSSION THREAD

**Discussion by murrant** [composer/composer#10362] [Q&A / Support]
I work on an open source project that users install on their local servers.
We include composer.json and composer.lock in the repo to make sure installs are running the correct versions.
An automatic update script runs composer install to make sure dependencies are updated when the code is.

We want to allow users to install additional packages, but the update script will either remove the local changes and packages or the local changes will block auto updates of the composer.json + composer.lock when we update them upstream.

Are there any good solutions for this use case?

**Comment by GromNaN** [comment]
Instead of allowing changes to the `composer.json`, extra packages could be stored in an other file. You create a script in your update process that run `composer require <package>` for each. Composer will not update other packages.

**Reply by murrant** [reply]
So for every update:
1. reset composer.json and composer.lock to upstream
2. update code
3. require the added packages
4. call composer install
 
Is that what you are suggesting?

**Reply by GromNaN** [reply]
Yes. That was the idea. But there is a better solution.

You don't have to ship the composer.lock to lock dependencies. You can create a meta package (no code inside) with fixed requirements (exact version numbers). This package will only add constraints.


```
{
    "vendor/dependencies",
    "require": {
        "monolog/monolog": "1.0.17"
    }
}  
```
